### ISSUE TITLE
xs3_filter_fir_s32 saturation issue

### ISSUE THREAD

**Issue by keithm-xmos** [xmos/lib_xcore_math#119]
The output of `xs3_filter_fir_s32` may be `INT32_MIN` instead of `INT32_MAX` when the new sample equals `INT32_MAX`. 

Filter coeffs are: 

```
const int32_t WORD_ALIGNED fir_bypass_coefs[TAP_COUNT_fir_bypass] = {
  -0x80000000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 
	0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 
};
```

Input signal is a sinusoidal wav in the range `[INT32_MIN, INT32_MAX]` inclusive.

**Comment by keithm-xmos** [comment]
It appears that the biquad filter function suffers from the same bug.  This has not been addressed in https://github.com/xmos/lib_xcore_math/pull/120

We may want to add a regression test for this in addition to implementing the fixes.

**Comment by astewart-xmos** [comment]
The biquad filter appears to be functioning as intended.  The issue here is that all 32-bit internal filter states (and including outputs) have a **fixed-scale** equal to that of the input; they are not **fixed-precision**.  

The output of each filter section is computed as the inner product of that section's filter state and the corresponding coefficients. This inner product is stored internally in a 40-bit accumulator which, because coefficients can only span -2.0 to 2.0, cannot possibly ever saturate.  However, the accumulator _can_ go beyond the `[INT32_MIN, INT32_MAX]` range.  The accumulator is not saturated to 32-bit bounds when the output of one filter section (which is also the input to the next!) is stored back into the filter state.  If the accumulator was beyond the 32-bit range, an overflow occurs and the state gets corrupted.

In the FIR filter case, the filter is saturated once at the very end, and this is deemed to be acceptable -- partly because it is at least _detectable_.  In order to avoid this overflow issue with the biquad filters, a bunch of additional instructions would be required. But more importantly, _if the saturation logic actually **did** anything_, it would almost certainly be an unacceptable corruption of the filter state, possibly even driving the filter to become unstable.

All the [python script](https://github.com/xmos/lib_xcore_math/blob/develop/lib_xcore_math/script/gen_biquad_filter_s32.py) (which generates biquad filters from floating-point coefficients) does with the coefficients is to convert them directly to Q1.30 format.  Trying to have the script analyze the filter to determine whether overflow is possible is beyond the scope of what it should be expected to do.

However, I think it makes sense to include an optional floating-point parameter in the script by which all the coefficients will be scaled.  If the filter does appear to suffer from overflow, scaling down the coefficients may avoid it.  Unfortunately, it will not always be immediately obvious whether _some_ sequence of inputs might cause saturation.

Alternatively the user can try scaling the input.

In either case, the user would have to be aware of the scaling and deal with the output accordingly.

When converting an existing filter with existing test inputs, one useful heuristic is that the outputs of individual filter sections should never grow larger in magnitude than the largest magnitude input to the filter. Because otherwise if the input is scaled to the 32-bit range, the filter states will grow beyond the 32-bit range.

**Comment by keithm-xmos** [comment]
@astewart-xmos  Question about the original fir_bypass_coefs case... 

The output of this filter appears to be the -1*input (see attached image).  Is this expected given the coefficients:   

```
const int32_t WORD_ALIGNED fir_bypass_coefs[TAP_COUNT_fir_bypass] = {
  -0x80000000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 
	0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 
};
```

<img width="812" alt="Screen Shot 2022-11-01 at 6 44 50 PM" src="https://user-images.githubusercontent.com/35048036/199356194-d722a856-0469-436d-a1cc-4c740a09bdcd.png">

(mentioning @QuinnWang so he can follow this thread)

**Comment by QuinnWang** [comment]
The fir bypass coeffs should be 1,0,0,0..., and the Q1.30 format should be +0x80000000, 0, 0, 0...? The **-0x80000000**, 0, 0, 0... one used in the fir_bypass.c was generated by the python script gen_fir_filter_s32.py, so looks the gen_fir_filter_s32.py problem? @astewart-xmos

**Comment by QuinnWang** [comment]
I just tried the coeffs in the .csv to be -1, 0, 0..., and the gen_fir_filter_s32.py generated -0x80000000, 0, 0... as well... so should be the script bug already

**Comment by QuinnWang** [comment]
I just tried the coeffs in the .csv to be -1, 0, 0..., and the gen_fir_filter_s32.py generated -0x80000000, 0, 0... as well... so should be the script bug already

**Comment by QuinnWang** [comment]
But, when I tried to change the -0x80000000 to +0x80000000 in the fir_bypass.c directly, the processed audio wave file looks still turning upside down. A bit confusion.

**Comment by astewart-xmos** [comment]
For the 32-bit FIR filter, the filter behavior is determined by both the coefficients _and_ the [output shift](https://github.com/xmos/lib_xcore_math/blob/master/lib_xcore_math/api/xmath/filter.h#L264). 

@keithm-xmos In the case you mention (_assuming_ that Audacity isn't doing something funky when it converts from PCM samples to float samples (an assumption I wouldn't stake my life on)), iff the output shift is `1`, then yes, that's exactly the result I would expect.

> The fir bypass coeffs should be 1,0,0,0..., and the Q1.30 format should be +0x80000000, 0, 0, 0...?

@QuinnWang The value 1.0 in Q1.30 is  `0x40000000` (2**30), not `0x80000000` (2**31). Are you saying a positive coefficient went in but a negative coefficient came out? That should never happen.

> I just tried the coeffs in the .csv to be -1, 0, 0..., and the gen_fir_filter_s32.py generated -0x80000000, 0, 0... as well... so should be the script bug already

@QuinnWang What was the generated value for the [output shift](https://github.com/xmos/lib_xcore_math/blob/master/lib_xcore_math/api/xmath/filter.h#L264)?  Iff it was `1`, then this is as expected, because `-0x80000000` = -2.0 in Q1.30 will get multiplied by the input, but then an output shift of `1` would effectively reduce that to `-1.0`.

> But, when I tried to change the -0x80000000 to +0x80000000 in the fir_bypass.c directly, the processed audio wave file looks still turning upside down. A bit confusion.

`+0x80000000` is not a representable value in a 32-bit signed integer `INT32_MAX == +0x7FFFFFFF`. Adding `1` to `INT32_MAX` gives `-0x80000000 == INT32_MIN`.  So, changing from `-0x80000000` to `+0x80000000` should not change the behavior because the literal `+0x80000000` assigned to an `int32_t` will become `-0x80000000`.

**Comment by QuinnWang** [comment]
> @QuinnWang The value 1.0 in Q1.30 is 0x40000000 (2^30), not 0x80000000 (2^31). Are you saying a positive coefficient went in but a negative coefficient came out? That should never happen.

Hi @astewart-xmos , for the fir bypass goal, we need to 1,0,0..., so the 1st coeff should be 0x40000000?! Then the script should generate 0x40000000 instead of current -0x80000000. This does happen, since I was using the 1,0,0... in the .csv to generate the coeffs... Could you please try to see if you can reproduce it? Thanks

**Comment by astewart-xmos** [comment]
@QuinnWang  You're correct -- that is a bug in the python script. I reproduced it locally and will issue a fix for it.

I'm a little embarrassed this simple case slipped by me.  Thanks for finding it!

**Comment by astewart-xmos** [comment]
@QuinnWang  I've pushed a fix for the FIR filter conversion python scripts into the `fix/119` [branch](https://github.com/xmos/lib_xcore_math/tree/fix/119).  A new release should be issued soon containing these fixes, but you can find the fixes there immediately.

The issue you found was also in the 16-bit FIR conversion script.

In both scripts the bug is a corner case that only occurs when the largest-magnitude filter coefficient happens to be a positive power of 2.  In that case, the coefficients get scaled up such that that maximum coefficient is `0x80000000`, which, interpreted as an `int32_t` is actually the value `-0x80000000`.

**Comment by astewart-xmos** [comment]
I also noticed another issue while updating the 32-bit FIR script where the output exponent value generated by the script (e.g. if the filter name is `FilterName`, then the output exponent value is called `FilterName_exp`) has a confusing interpretation.

In that case, `FilterName_exp` is the exponent associated with the output values **only if** the exponent associated with the filter _input_  is `-30`.  (That wasn't the original intention, but that is the correct interpretation of what it actually produces).

But FIR filters are linear, so it should be completely agnostic to the exponent associated with the input.  To fix this confusion I've updated the generated header to declare the `FilterName_exp` as **deprecated**.  I've added in another output value which would be called `FilterName_exp_diff`, which is the _change_ in exponent from the input to the output of the filter. 

I did the same thing with the 16-bit FIR just to be consistent, although the original `FilterName_exp` for 16-bit already had the same interpretation as `FilterName_exp_diff`.

----

For the sake of clarity, I'll give an example here:

```c
  #include <stdio.h>
  #include <math.h>
  #include "FilterName.h"
  
  // Process one sample with the filter, but print out the floating-point 
  // equivalent value of the input and output samples
  int32_t foo(int32_t sample_in){
    // assume already initialized
  
    // Exponent associated with input sample.
    // Note: This must not change without clearing the filter state.
    const exponent_t input_exp = -22;

    // print floating-point input sample value
    printf("sample_in = %f\n", ldexpf(sample_in, input_exp));
    
    // apply filter
    int32_t sample_out = FilterName(sample_in);

    // The exponent associated with the output sample
    const exponent output_exp = input_exp + FilterName_exp_diff;

    printf("sample_out = %f\n", ldexpf(sample_out, output_exp));

    return sample_out;
  }
```

Note that neglecting to take `FilterName_exp_diff` into account with the output samples is equivalent to applying a gain of `2**(-FilterName_exp_diff)` to the filter.

**Comment by astewart-xmos** [comment]
These issues have been addressed in PR #120 and fixes should be available in `v2.1.1`
